### PR TITLE
Added in conversations!

### PR DESCRIPTION
**Pull request by b1gd3vd0g**
This is a really big update guys! there were a lot of things i needed to change/add for dialog.

NOTES: it is not a complete system! I want to get a version of it out so that the pull request isn't completely unmanageable.
Conversations involving choices need to be made possible still.

In addition to dialog, I also have made the GameEngine into the 3 grounds (back, mid, fore). It draws on midground by default.

### CODE REVIEW

File: js/classes/conversation.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+class Conversation {
+    /**
+     * 
+     * @param {Array<DialogBubble>} array The array containing all possible
```
Comment (tclaar) [inline js/classes/conversation.js]: I'm sure you plan to continue working on the JSDoc in this file but don't forget to finish this sentence. :)

File: js/classes/conversation.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        this.array = array;
+        this.curr = 0;
+        this.new = isNew;
+    };
+
+    /**
+     * As the conversation is an entity, it must have an update loop. The update really just
+     * listens for user input, and will change the dialog bubble shown accordingly.
+     */
+    update() {
```
Comment (tclaar) [inline js/classes/conversation.js:21]: I appreciate all the comments in here!

File: js/classes/conversation.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        this.removeFromWorld = false;
+        this.curr = 0;
+        this.array.forEach((bubble) => {
+            bubble.refresh();
+        });
+        // Add it to the game.
+        GAME.addEntity(this);
+        // Set up the game for dialog listeners.
+        GAME.mode = GameEngine.DIALOG_MODE;
+        GAME.configureEventListeners();
```
Comment (tclaar) [inline js/classes/conversation.js:77]: Instead of setting `GAME.mode` and then calling `configureEventListeners`, I feel like it may be cleaner to allow the mode to be passed in as a parameter to that method and set `mode` within it.

File: js/classes/entities/midground/chad.js [hunk 4]
```diff
@@ -318,6 +321,11 @@ class Chad {
                         ZONE = entity.target;
                         ZONE.load();
                     }
+                    else if (entity.conversation) {
+                        if (GAME.user.interacting) {
```
Comment (tclaar) [inline js/classes/entities/midground/chad.js:325]: Not important at all, but you could probably combine these into a single if condition if you won't be adding other branches.

File: js/classes/conversation.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    } else {
+                        // Move onto the next bubble.
+                        this.curr++;
+                        GAME.addEntity(this.array[this.curr]);
+                    }
+                }
+            } else {
+                // currentBubble has not finished its typing.
+                // player wants to see currentBubble complete now.
+                this.array[this.curr].finishTyping();
```
Comment (NathanHinthorne) [inline js/classes/conversation.js:50]: Nice idea with this asynchronous task!

File: js/classes/conversation.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * but really it is a container/handler for the dialog bubbles in its array.
+ * Conversations can be stored in entities, but when they are active, the game goes into Dialog mode.
+ */
+class Conversation {
+    /**
+     * @param {Array<DialogBubble>} array The array containing all possible dialog bubbles in this conversation.
+     * @param {boolean} isNew Is it Chad's first time seeing this conversation? Default: true.
+     */
+    constructor(array, isNew = true) {
+        this.array = array;
```
Comment (calebkrauter) [inline js/classes/conversation.js:12]: I would suggest that we name the array after what it contains. It is hard to follow.

**Review by NathanHinthorne — APPROVED** [review]
Thanks for the much-needed dialog changes @B1GD3VD0G !

File: js/classes/conversation.js (hunk 5 above)
Comment (calebkrauter) [reply js/classes/conversation.js:50]: Love this.

File: js/classes/conversation.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+     * listens for user input, and will change the dialog bubble shown accordingly.
+     */
+    update() {
+        // Is the user wanting to move on to the next dialog bubble?
+        if (GAME.user.continuingConversation) {
+            // The user has pressed space.
+
+            const currentBubble = this.array[this.curr];
+
+            if (currentBubble.currentLine >= currentBubble.targetLines.length) {
```
Comment (calebkrauter) [inline js/classes/conversation.js:28]: This class looks great however I am confused on how the typing is actually displayed one letter at a time. How does this work? @B1GD3VD0G

File: js/classes/entities/midground/chad.js [hunk 8]
```diff
@@ -271,7 +274,7 @@ class Chad {
         this.isOnGround = false;
 
         // Step 4: Have we collided with anything?
-        GAME.entities.forEach((entity) => {
+        GAME.entities.midground.forEach((entity) => {
```
Comment (calebkrauter) [inline js/classes/entities/midground/chad.js:277]: Nice job with optimizing the different layering of the world.

File: js/classes/entities/midground/chad.js [hunk 9]
```diff
@@ -318,6 +321,11 @@ class Chad {
                         ZONE = entity.target;
                         ZONE.load();
                     }
+                    else if (entity.conversation) {
```
Comment (calebkrauter) [inline js/classes/entities/midground/chad.js:324]: Very intuitive. Love it.

File: js/classes/entities/midground/npc/village/papa_chad.js [hunk 10]
```diff
@@ -71,7 +70,7 @@ class PapaChad {
 
 
         // Step 4: Have we collided with anything?
-        GAME.entities.forEach((entity) => {
+        GAME.entities.midground.forEach((entity) => {
```
Comment (calebkrauter) [inline js/classes/entities/midground/npc/village/papa_chad.js:73]: Could this just be an instance of a class with a simple parameter pass? Might make it slightly cleaner and allow us to just reuse it for NPCs. Minor but it might improve it a bit.

File: js/classes/entities/midground/weapon/projectile.js [hunk 11]
```diff
@@ -165,7 +165,7 @@ class Projectile {
             this.pos = Vector.add(this.pos, new Vector(posChange.x, posChange.y + this.yVelocity))
 
             this.updateBoundingBox();
-            GAME.entities.forEach((entity) => {
+            GAME.entities.midground.forEach((entity) => {
                 if (this != entity && entity.boundingBox) {
```
Comment (calebkrauter) [inline js/classes/entities/midground/weapon/projectile.js:179]: In sword.js we use bb for bounding box and here we use boundingBox. Can we choose one for naming bounding boxes? Also can we put this code in a class where it can be instantiated and the code doesn't have to be copied in multiple classes to clean things up? I see this happening in sword.js and imagine we may use some of this code in other classes for NPCs or similar.

File: js/classes/game_engine.js [hunk 12]
```diff
@@ -35,26 +42,39 @@ class GameEngine {
 
         /** The timer tells you how long it's been since the last tick! */
         this.timer = new Timer();
-        /** Are we currently debugging? */
-        this.debug = false;
+        
+        this.conversation = null;
+
+        this.mode = GameEngine.GAMEPLAY_MODE;
```
Comment (calebkrauter) [inline js/classes/game_engine.js:48]: What other modes are there?

File: js/classes/game_engine.js [hunk 13]
```diff
@@ -35,26 +42,39 @@ class GameEngine {
...
     };
 
     /**
      * This adds a new entity to the entities array.
      * @param {Object} entity The entity (sprite) that you want to add to the Game.
+     * @param {number} zIndex negative for background, zero (default) for midground, positive for foreground.
      */
-    addEntity(entity) {
-        this.entities.push(entity);
+    addEntity(entity, zIndex = 0) {
```
Comment (calebkrauter) [inline js/classes/game_engine.js:56]: Nice work on this!

File: js/classes/game_engine.js [hunk 14]
```diff
@@ -74,25 +94,42 @@ class GameEngine {
      * This method is going to go through all entities and allow them to update their position.
      */
     update() {
-        // Iterate forward through the entities array. Update everything.
-        for (let i = 0; i < this.entities.length; i++) {
-            let entity = this.entities[i];
-            if (!entity.removeFromWorld) {
+        
+        // (1) Update the background entities:
+        this.entities.background.forEach((entity) => {
```
Comment (calebkrauter) [inline js/classes/game_engine.js]: We may only be doing this three times for foreground, mid-ground and background however we could possibly improve this by reusing the same method as it appears we are copying quite a bit of code.

File: js/classes/game_engine.js [hunk 15]
```diff
@@ -102,105 +139,56 @@ class GameEngine {
...
      * This method is going to start listening for user inputs. Affects the Game Engine's left/right/up/down booleans according
      * to interaction with either the WASD keys or arrows.
      */
-    startInput() {
-
-        CANVAS.addEventListener("mousedown", (e) => {
-            this.user.firing = false;
-            this.user.aiming = true;
+    configureEventListeners() {
+        Object.keys(GAME.user).forEach((key) => {
```
Comment (calebkrauter) [inline js/classes/game_engine.js:174]: Great work cleaning this up!

File: js/classes/game_engine.js [hunk 16]
```diff
@@ -102,105 +139,56 @@ class GameEngine {
...
+        if (this.mode === GameEngine.GAMEPLAY_MODE) {
+            // Remove all other listeners.
+            CANVAS.removeEventListener("keypress", EVENT_HANDLERS.dialogKeyPress, false);
+            // Add the gameplay listeners.
+            CANVAS.addEventListener("mousedown", EVENT_HANDLERS.gameplayMouseDown, false);
+            CANVAS.addEventListener("mouseup", EVENT_HANDLERS.gameplayMouseUp, false);
+            CANVAS.addEventListener("mousemove", EVENT_HANDLERS.gameplayMouseMove, false);
+            CANVAS.addEventListener("keydown", EVENT_HANDLERS.gameplayKeyDown, false);
+            CANVAS.addEventListener("keyup", EVENT_HANDLERS.gameplayKeyUp, false);
+        } else if (this.mode === GameEngine.DIALOG_MODE) {
```
Comment (calebkrauter) [inline js/classes/game_engine.js:187]: Good idea on how to effectively pause controls.

File: js/scripts/conversation_loaders/playground_conversation_loaders.js [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            // 3
+            new DialogBubble(
+                chad,
+                "Ok dad sounds good!",
+                true)
+        ],
+        // The following works perfectly - 1 person talking just going from line to line.
+        jerseyShore: [
+            new DialogBubble(
+                papa,
```
Comment (calebkrauter) [inline js/scripts/conversation_loaders/playground_conversation_loaders.js:39]: Lol

File: js/scripts/util.js [hunk 18]
```diff
@@ -15,10 +15,18 @@ window.requestAnimFrame = (() => {
         });
 })();
 
+/* This fixes the bug where the window scrolls down when you press space. */
+window.onkeydown = (e) => {
```
Comment (calebkrauter) [inline js/scripts/util.js:19]: Wonderful!

File: js/scripts/util.js [hunk 19]
```diff
@@ -133,4 +140,86 @@ const checkBlockCollisions = (entity) => {
     entity.boundingBox = new BoundingBox(entity.pos, entity.constructor.SCALED_SIZE);
 
     return collisions;
+};
+
+// The following is necessary because we must change the listeners for different modes (right now, gameplay and dialog).
+/** Contains all functions called as event handlers. */
+const EVENT_HANDLERS = {
```
Comment (calebkrauter) [inline js/scripts/util.js:148]: I like how clean this approach is very nice!

File: js/scripts/util.js [hunk 20]
```diff
@@ -133,4 +140,86 @@ const checkBlockCollisions = (entity) => {
...
+    gameplayMouseDown: (e) => {
+        GAME.user.firing = false;
+        GAME.user.aiming = true;
+    },
+    gameplayMouseUp: (e) => {
+        GAME.user.aiming = false;
+        GAME.user.firing = true;
+    },
+    gameplayMouseMove: (e) => {
+        const rect = CANVAS.getBoundingClientRect();
```
Comment (calebkrauter) [inline js/scripts/util.js:158]: What does CANVAS.getBoundingClientRect do? is that the bounding box for Chad? Can you explain?

File: js/scripts/util.js [hunk 21]
```diff
@@ -133,4 +140,86 @@ const checkBlockCollisions = (entity) => {
...
+        const scaleX = CANVAS.width / rect.width;
+        const scaleY = CANVAS.height / rect.height;
+        GAME.mousePos = new Vector((e.clientX - rect.left) * scaleX, (e.clientY - rect.top) * scaleY);
+    },
+    gameplayKeyDown: (e) => {
```
Comment (calebkrauter) [inline js/scripts/util.js:163]: Similar to what we have in some GameEngine code we could likely put each of our keys into an array and loop over it in a for loop to clean it up. Not necessary but might be nice to clean up more. Also why are we doing this in Util? Can you explain that?

File: js/classes/entities/foreground/dialog_bubble.js [hunk 22]
```diff
@@ -34,6 +42,164 @@ class DialogBubble {
...
+            );
+        this.speaker.drawFrame(speakerStart, scale);
+
+        // (3) Draw the text.
+        let textStart = new Vector(
+            (Camera.SIZE.x - DialogBubble.SCALED_SIZE.x) / 2 + (0.1 * DialogBubble.SCALED_SIZE.x),
+            Camera.SIZE.y - DialogBubble.SCALED_SIZE.y + (0.1 * DialogBubble.SCALED_SIZE.x) - (5 * DialogBubble.SCALE) + 17
+        );
+        // The jump in pixels after each line!
+        const textJump = new Vector(0, 40); // NOTE: this is based off a text size of 34!
```
Comment (tclaar) [inline js/classes/entities/foreground/dialog_bubble.js:97]: Maybe make this a static constant as well?

File: js/classes/entities/foreground/dialog_bubble.js [hunk 23]
```diff
@@ -34,6 +42,164 @@ class DialogBubble {
...
+        // (1) Draw the DialogBubble sprite.
+        CTX.drawImage(
+            ASSET_MGR.getAsset(DialogBubble.SPRITESHEET),
+            0, DialogBubble.SIZE.y * this.type,
+            DialogBubble.SIZE.x, DialogBubble.SIZE.y,
+            dbPos.x, dbPos.y,
+            DialogBubble.SCALED_SIZE.x, DialogBubble.SCALED_SIZE.y);
+            
+        // (2) Draw the speaker.
+        const scale = 6;
```
Comment (tclaar) [inline js/classes/entities/foreground/dialog_bubble.js:80]: Maybe make this a static constant?

File: js/classes/entities/foreground/dialog_bubble.js [hunk 24]
```diff
@@ -34,6 +42,164 @@ class DialogBubble {
...
+            ) : new Vector(
+                Camera.SIZE.x - ((Camera.SIZE.x - DialogBubble.SCALED_SIZE.x) / 2) + (this.speaker.size.x * 0.5),
+                Camera.SIZE.y - DialogBubble.SCALED_SIZE.y
+            );
+        this.speaker.drawFrame(speakerStart, scale);
+
+        // (3) Draw the text.
+        let textStart = new Vector(
+            (Camera.SIZE.x - DialogBubble.SCALED_SIZE.x) / 2 + (0.1 * DialogBubble.SCALED_SIZE.x),
+            Camera.SIZE.y - DialogBubble.SCALED_SIZE.y + (0.1 * DialogBubble.SCALED_SIZE.x) - (5 * DialogBubble.SCALE) + 17
```
Comment (tclaar) [inline js/classes/entities/foreground/dialog_bubble.js:94]: Depending on what some of these number literals (0.1, 5, 17) represent, maybe they could be converted into static constants.

**Review by calebkrauter — COMMENTED** [review]
Looks fantastic!

File: js/classes/entities/foreground/dialog_bubble.js [hunk 25]
```diff
@@ -34,6 +42,164 @@ class DialogBubble {
...
+                space = " ";
+                wordIndex++;
+                if (wordIndex === words.length) break;
+            }
+            // Now, the next word cannot fit on this line.
+            // Append currLine to lines.
+            lines.push(currLine);
+            currLine = "";
+        }
+        if (lines.length > 6) throw new Error("Cannot fit the following string into a single dialog bubble: " + text);
```
Comment (tclaar) [inline js/classes/entities/foreground/dialog_bubble.js:162]: Maybe make a MAX_LINES constant? I'll stop suggesting static constants now sorry lol.

File: js/classes/conversation.js (hunk 6 above)
Comment (b1gd3vd0g) [reply js/classes/conversation.js:12]: i feel like Conversation.array is kinda self-explanatory; a conversation array is a term already established in the conversation loaders. dialogBubbleArray is too wordy and IMO does not improve clarity.

File: js/classes/entities/midground/npc/village/papa_chad.js [hunk 26]
```diff
@@ -5,7 +5,11 @@
  * @author Devin, Caleb, Nathan, Trae
  */
 class PapaChad {
-    constructor(pos) {
+    /**
+     * @param {Vector} pos the position at which he should spawn. 
+     * @param {Conversation} convo The conversation that will show if PapaChad is interacted with.
+     */
+    constructor(pos, convo = null) {
```
Comment (tclaar) [inline js/classes/entities/midground/npc/village/papa_chad.js:12]: Good call passing in `convo` as a parameter instead of hardcoding the converation(s) into the class itself.

File: js/classes/conversation.js (hunk 7 above)
Comment (b1gd3vd0g) [reply js/classes/conversation.js:28]: There are 3 instance fields of any DialogBubble (let's call an instance db) you should be familiar with: db.targetLines (contains all text, finished on construction), db.visibleLines (the text that's actually shown), and db.currentLine. db.currentLine tells us which line of text we are currently adding to - if it is equal to visibleLines (or targetLines).length, we know we have finished writing completely. Otherwise, as a DialogBubble is an entity, every single update loop it will add a character to db.currentLine, if that line is finished it will increment db.currentLine.

File: js/classes/game_engine.js [hunk 27]
```diff
@@ -74,31 +94,46 @@ class GameEngine {
...
-            // Iterate forward through the entities array. Update everything.
-            for (let i = 0; i < this.entities.length; i++) {
-                let entity = this.entities[i];
-                if (!entity.removeFromWorld) {
+            // (1) Update the background entities:
+            this.entities.background.forEach((entity) => {
+                if (entity.removeFromWorld) {
+                    const i = this.entities.background.indexOf(entity);
+                    this.entities.background.splice(i, 1);
+                } else {
```
Comment (tclaar) [inline js/classes/game_engine.js:103]: I see why you refactored it this way but the only problem is that it allows entities to hang around for an extra update after setting the removeFromWorld flag. Not likely to matter but I thought I'd mention it.

File: js/classes/game_engine.js (hunk 12 above)
Comment (b1gd3vd0g) [reply js/classes/game_engine.js:48]: gameplay and dialog. I predict there will be a menu mode, but there's not yet.

File: js/classes/game_engine.js [hunk 28]
```diff
@@ -74,31 +94,46 @@ class GameEngine {
...
             CHAD.update();
-            // Update the camera, which is not a regular entity.
             CAMERA.update();
 
-            // Iterate backward through the entities array, and remove all entities which ought be removed.
-            for (let i = this.entities.length - 1; i >= 0; i--) {
-                if (this.entities[i].removeFromWorld) {
-                    this.entities.splice(i, 1);
+            // (4) Update the foreground entities:
+            this.entities.foreground.forEach((entity) => {
```
Comment (tclaar) [inline js/classes/game_engine.js:123]: We could probably generalize these forEach loops into a function that accepts the fore/mid/background array as a parameter to simplify this code.

File: js/classes/game_engine.js (hunk 14 above)
Comment (b1gd3vd0g) [reply js/classes/game_engine.js]: Only in the game engine itself, and only inside a single method. I think we're ok.

File: js/classes/game_engine.js [hunk 29]
```diff
@@ -132,92 +170,32 @@ class GameEngine {
...
-        });
-
-        CANVAS.addEventListener("mouseup", (e) => {
-            if (e.button === 0) {
-                this.user.aiming = false;
-                this.user.firing = true;
-            } else if (e.button === 2) {
-                // TODO update user.jabbing here
-            }
+    configureEventListeners() {
```
Comment (tclaar) [inline js/classes/game_engine.js:173]: I like this method name much better than the old one.

File: js/classes/game_engine.js (hunk 15 above)
Comment (b1gd3vd0g) [reply js/classes/game_engine.js:174]: Not just cleaning; this was necessary in order to be able to reconfigure the CANVAS's listeners without reinitializing it and the context (which might have been the way to go, but screw it, what's done is done)

File: js/scripts/conversation_loaders/playground_conversation_loaders.js (hunk 17 above)
Comment (b1gd3vd0g) [reply js/scripts/conversation_loaders/playground_conversation_loaders.js:39]: Finally someone laughs lmao

File: js/scripts/util.js (hunk 20 above)
Comment (b1gd3vd0g) [reply js/scripts/util.js:158]: Ask @NathanHinthorne

File: js/scripts/util.js [hunk 30]
```diff
@@ -127,4 +141,86 @@ const checkBlockCollisions = (entity) => {
     entity.boundingBox = new BoundingBox(entity.pos, entity.constructor.SCALED_SIZE);
 
     return collisions;
+};
+
+// The following is necessary because we must change the listeners for different modes (right now, gameplay and dialog).
+/** Contains all functions called as event handlers. */
+const EVENT_HANDLERS = {
```
Comment (tclaar) [inline js/scripts/util.js:148]: I'm not certain this belongs in util, as it seems like it is relevant only to the game engine. I understand game engine is cluttered enough as is but I'm not sure this is the right place.

File: js/scripts/util.js (hunk 21 above)
Comment (b1gd3vd0g) [reply js/scripts/util.js:163]: These functions are necessary in util because we must define the event handlers separately, because they are necessary for the "removeEventHandler" function call. Simply copy pasting or similar techniques do not work, because it would not actually be the same function (same place in memory), just a function which does the same thing.

**Review by tclaar — APPROVED** [review]
Great stuff Devin! This is really well done. I have a few comments but nothing major - you did a very good job.

File: js/classes/conversation.js (hunk 3 above)
Comment (b1gd3vd0g) [reply js/classes/conversation.js:77]: I don't disagree. Not sure how important that will be, but I appreciate the feedback.

File: js/classes/conversation.js (hunk 6 above)
Comment (calebkrauter) [reply js/classes/conversation.js:12]: I would suggest changing it to bubbles or dialog and the fact that it is an array doesn't need to be specified as it is clear when you use it. Conversation.array does work.

File: js/scripts/util.js (hunk 20 above)
Comment (NathanHinthorne) [reply js/scripts/util.js:158]: There's a lot of details in getBoundingClientRect(). It's based on stuff related to the canvas the user sees. The function helps with the conversion process of mouse-to-canvas coordinates, ensuring the coordinates are scaled properly based on the user's screen size. I ran into a lot of bugs and decided to use this function to solve the problem.

Here's some more info on it:
https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect

File: js/classes/entities/midground/chad.js (hunk 4 above)
Comment (b1gd3vd0g) [reply js/classes/entities/midground/chad.js:325]: I like the cleanliness of having it separated, but that's true.

File: js/classes/conversation.js (hunk 7 above)
Comment (calebkrauter) [reply js/classes/conversation.js:28]: Gotcha, thank you!

File: js/classes/entities/foreground/dialog_bubble.js (hunk 22 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:97]: I could, but it's only relevant in a single method so wouldn't really be necessary I think.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 23 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:80]: Being honest, right now I'm only drawing papa chad. scale could be different for different speakers. So maybe static won't event be the way to go. I will do more work on this this week.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 24 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:94]: Ehh, maybe const variables but I don't think static constants would be appropriate here.

File: js/classes/entities/foreground/dialog_bubble.js (hunk 25 above)
Comment (b1gd3vd0g) [reply js/classes/entities/foreground/dialog_bubble.js:162]: This one is definitely a good idea. Thanks!

File: js/classes/entities/midground/npc/village/papa_chad.js (hunk 26 above)
Comment (b1gd3vd0g) [reply js/classes/entities/midground/npc/village/papa_chad.js:12]: Yeah, I mean depending on our story progress his convo would be different so we gotta be able to determine that.

File: js/classes/game_engine.js (hunk 27 above)
Comment (b1gd3vd0g) [reply js/classes/game_engine.js:103]: I don't think I changed anything Order-of-Operations wise for removal. The removeFromWorld flag is set, and then its next update method it will be removed. Actually removing an entity immediately is much more verbose, and not really using the code wisely IMO.

File: js/classes/game_engine.js (hunk 29 above)
Comment (b1gd3vd0g) [reply js/classes/game_engine.js:173]: Me too. startInput() was from prof's code lol. I never liked it.

File: js/classes/game_engine.js (hunk 28 above)
Comment (b1gd3vd0g) [reply js/classes/game_engine.js:123]: Seems like a popular opinion. I'll look into it.

File: js/scripts/util.js (hunk 21 above)
Comment (calebkrauter) [reply js/scripts/util.js:163]: Okay now it makes sense why these functions are in util. I see what you mean about the cleaning up of the code, I suppose if we were to do that it would not actually simplify it by using a loop.

File: js/scripts/util.js (hunk 30 above)
Comment (b1gd3vd0g) [reply js/scripts/util.js:148]: I don't think I agree with you. Realistically, it has little to do with the game engine, it's only that the game engine is what's going to be calling these. These would be in the Canvas class (if we had one) but I think that util is okay. In the html comment above the util import, it says "Scripts containing info/functions for classes" which is exactly what this is.

File: js/scripts/util.js (hunk 20 above)
Comment (calebkrauter) [reply js/scripts/util.js:158]: Thank you!
